### ISSUE TITLE
Message log service

### ISSUE THREAD

**Issue by mjgpy3** [PolyglotSymposium/void#42]
@Kazark, what if we had a service (primarily for debugging) that logged out messages (events and commands)?

Thoughts?

I think it would only be worth it if it were smart enough to "catch-all" (at least for the moment).

**Comment by Kazark** [comment]
Yeah I've toyed with the idea. In theory we should be able to subscribe an abstract `Message` channel and receive them all, and in theory we could get them printed using `printf "%A"`.

The flip side to something like this is, if you need it, you have to ask yourself if there is something else about your system that is not easy to understand, and if that's why you need it. Not saying it wouldn't be useful, just saying that if there is pain, I want to feel what that pain is, and maybe there's another way to solve it.

Also, we would want to make sure that it would be something you could turn on and off.

Moreover, it seems that it would be most useful if it told you _who_ published or received what. On the other hand, maybe some static analysis would be better for this. Or some kind of thing that, if a command does not get handled, sends out a `notImplemented` message.

In sum, this could be a good idea, but I'd like to discuss why exactly we need it before jumping on it. Maybe there is something else that we need instead. Or maybe we can clarify what exactly it is that we need from this.

**Comment by mjgpy3** [comment]
> Also, we would want to make sure that it would be something you could turn on and off.

Totally agree

> The flip side to something like this is, if you need it, you have to ask yourself if there is something else about your system that is not easy to understand, and if that's why you need it. Not saying it wouldn't be useful, just saying that if there is pain, I want to feel what that pain is, and maybe there's another way to solve it.

Right. I'm thinking primarily as a new developer to the system. You explained the architecture well, but a concrete example is worth a lot.

Also, I believe this would help alleviate the "I forgot to register to the bus" error because it would be more clear where messages _are not_ being published/handled.

**Comment by mjgpy3** [comment]
Also it might be useful as a tool to see if we've gotten our abstractions right. I.e. helping to answer questions like _do the events (or subscribers or publishers) make sense as is?_.

**Comment by Kazark** [comment]
Okay, so if I can summarize what you are saying:
1. Easier to grasp the system
   - Particularly for a developer new to the system.
   - For an overall, abstract view of the system to help us analyze it from a bird's eye view
2. Even with the new system you can still forget to register

Any other problems that we are trying to solve with this suggestion, or that are related but this suggestion may not solve?

**Comment by mjgpy3** [comment]
Spot-on summary, thanks. No other things come to mind.

**Comment by Kazark** [comment]
Okay, with respect to 2, I've been thinking: how are plugins written in F# or other .NET languages going to register their handlers? Basically, we will need to reflect on the assembly they they are contained in. But while we are doing that... couldn't we solve our own "Forget to register" problem?

What if we introduced a `[<AutoSubscribe>]` attribute for `Module.Service.subscribe` functions, and remove the centralized subscription entirely?

As for 1, I think that definitely needs to be addressed. I don't know if this is the best solution; however, it might be a solution that is worth a try. I wonder, though, if something with static analysis that could draw out the message dependencies of the system might actually be more useful than a runtime thing. The reason I say this is because I am afraid that once we actually build a message logger we may find that there is a lot of noise and it is hard to follow.

On other hand, you don't know until you try. At the very least, it's probably a valuable experiment.

**Comment by mjgpy3** [comment]
> What if we introduced a [<AutoSubscribe>] attribute for Module.Service.subscribe functions, and remove the centralized subscription entirely?

I had this thought as well. Sounds cool to me, it is probably a case of `Magic++`, but the benefits would be kinda awesome.

> I wonder, though, if something with static analysis that could draw out the message dependencies of the system might actually be more useful than a runtime thing

Very cool idea! Is there tooling in place to do such analysis?

> it's probably a valuable experiment.

Right. I saw this as being a small lift for potentially high benefit.

**Comment by mjgpy3** [comment]
"tooling in place" meaning "tooling that exists somewhere"

**Comment by Kazark** [comment]
Yeah, it is kind of magical, but it might be the right kind of magic. As long as it is still something explicit, i.e. you actually decorated your function, rather than just being convention based, like finding everything called `Service.subscribe(...)`. Also it just feels more explicit to bring all the registration back to the thing being registered.

As to the static analysis, yeah, there are a couple of .NET tools available. The names are popping off my tongue though. We would need to write some code for it, but there are some pretty solid libraries already in place.
